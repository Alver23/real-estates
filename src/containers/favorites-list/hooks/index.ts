// Dependencies
import { useMemo } from 'react';

// Hooks
import { useFetch } from '@alversoft/hooks/use-fetch';

// Services
import RealEstatesService from '@alversoft/services/real-estates';

// Interfaces
import { IRealEstateData } from '@alversoft/interfaces/real-estates';

export const useGetRealEstates = () => {
  const getList = () => {
    return RealEstatesService.getList();
  };

  const initialState = {
    data: {
      estates: [],
      included: undefined,
    },
  };
  const [state] = useFetch(getList, [], initialState);
  const {
    status,
    data: { estates, included },
  } = state;

  const newEstates = useMemo(() => {
    return estates.map((item: IRealEstateData) => {
      const { realEstateIds } = item;
      const estateDetailImages = realEstateIds
        .map((realEstateId) => {
          const { galleryUrls, name } = included[realEstateId] || { galleryUrls: [] };
          if (galleryUrls && galleryUrls.length > 0) {
            return {
              id: realEstateId.toString(),
              image: galleryUrls[0],
              alt: name,
            };
          }
          return undefined;
        })
        .filter((item) => !!item);
      return {
        ...item,
        images: estateDetailImages,
      };
    });
  }, [estates, included]);

  return [status, newEstates];
};

export default useGetRealEstates;
