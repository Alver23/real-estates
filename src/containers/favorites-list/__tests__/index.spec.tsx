// Dependencies
import { shallow, ShallowWrapper } from 'enzyme';

// Hooks
import * as RealEstatesHooks from '../hooks';

// Under test file
import FavoritesList from '../index';

describe('<FavoritesList />', () => {
  let component: ShallowWrapper;
  const mockState = ['idle', [{ id: 1, name: 'fake name', images: [] }]];

  beforeEach(() => {
    jest.spyOn(RealEstatesHooks, 'useGetRealEstates').mockReturnValue(mockState);
    component = shallow(<FavoritesList />);
  });

  it('should save a snapshot of the component', () => {
    expect(component).toMatchSnapshot();
  });
});
