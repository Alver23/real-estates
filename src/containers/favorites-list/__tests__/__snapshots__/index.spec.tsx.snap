// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FavoritesList /> should save a snapshot of the component 1`] = `
<div
  className="favority-container"
>
  <div
    className="favority-container--card-container"
    key="1"
  >
    <FavorityDetail
      images={Array []}
      name="fake name"
    />
  </div>
  <div
    className="favority-container--new-item"
  >
    <FavorityNewList />
  </div>
</div>
`;
